src/amms/erc_4626/mod.rs

```diff
@@ -0,0 +1,323 @@
+            U256,
+            U256,
+            U256,
+        )> as SolValue>::abi_decode(&res, false)?;
+        let (
+            vault_token,
+            vault_token_dec,
+            asset_token,
+            asset_token_dec,
+            vault_reserve,
+            asset_reserve,
+            deposit_fee_delta_1,
+            deposit_fee_delta_2,
+            deposit_no_fee,
+            withdraw_fee_delta_1,
+            withdraw_fee_delta_2,
+            withdraw_no_fee,
+        ) = if !data.is_empty() {
+            data[0]
+        } else {
+            todo!("Handle error")
+        };
+
+        // If both deltas are zero, the fee is zero
+        if deposit_fee_delta_1.is_zero() && deposit_fee_delta_2.is_zero() {
+            self.deposit_fee = 0;
+
+        // Assuming 18 decimals, if the delta of 1e20 is half the delta of 2e20, relative fee.
+        // Delta / (amount without fee / 10000) to give us the fee in basis points
+        } else if deposit_fee_delta_1 * U256_2 == deposit_fee_delta_2 {
+            self.deposit_fee = (deposit_fee_delta_1 / (deposit_no_fee / U256::from(10_000))).to();
+        } else {
+            todo!("Handle error")
+        }
+
+        // If both deltas are zero, the fee is zero
+        if withdraw_fee_delta_1.is_zero() && withdraw_fee_delta_2.is_zero() {
+            self.withdraw_fee = 0;
+        // Assuming 18 decimals, if the delta of 1e20 is half the delta of 2e20, relative fee.
+        // Delta / (amount without fee / 10000) to give us the fee in basis points
+        } else if withdraw_fee_delta_1 * U256::from(2) == withdraw_fee_delta_2 {
+            self.withdraw_fee =
+                (withdraw_fee_delta_1 / (withdraw_no_fee / U256::from(10_000))).to();
+        } else {
+            // If not a relative fee or zero, ignore vault
+            todo!("Handle error")
+        }
+
+        // if above does not error => populate the vault
+        self.vault_token = vault_token;
+        self.vault_token_decimals = vault_token_dec as u8;
+        self.asset_token = asset_token;
+        self.asset_token_decimals = asset_token_dec as u8;
+        self.vault_reserve = vault_reserve;
+        self.asset_reserve = asset_reserve;
+
+        Ok(self)
+    }
+}
+
+// TODO: swap calldata
+impl ERC4626Vault {
+    // Returns a new, unsynced ERC4626 vault
+    pub fn new(address: Address) -> Self {
+        Self {
+            vault_token: address,
+            ..Default::default()
+        }
+    }
+
+    pub fn get_amount_out(&self, amount_in: U256, reserve_in: U256, reserve_out: U256) -> U256 {
+        if amount_in.is_zero() {
+            return U256::ZERO;
+        }
+
+        if self.vault_reserve.is_zero() {
+            return amount_in;
+        }
+
+        let fee = if reserve_in == self.vault_reserve {
+            self.withdraw_fee
+        } else {
+            self.deposit_fee
+        };
+
+        amount_in * reserve_out / reserve_in * U256::from(10000 - fee) / U256_10000
```

Comment:
This should error if `fee == 10000` or `reserve_in == 0` with `Error::DivisionByZero`